data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/S3Source.java

```diff
@@ -47,10 +53,44 @@ public void start(Buffer<Record<Event>> buffer) {
         final ConfigBucketOwnerProviderFactory configBucketOwnerProviderFactory = new ConfigBucketOwnerProviderFactory();
         final BucketOwnerProvider bucketOwnerProvider = configBucketOwnerProviderFactory.createBucketOwnerProvider(s3SourceConfig);
 
-        final S3Service s3Service = new S3Service(s3SourceConfig, buffer, codec, pluginMetrics, bucketOwnerProvider);
+        Optional<S3SelectOptions> s3SelectOptional = Optional.ofNullable(s3SourceConfig.getS3SelectOptions());
+        S3ObjectPluginMetrics s3ObjectPluginMetrics = new S3ObjectPluginMetrics(pluginMetrics);
+
+        S3ClientBuilderFactory s3ClientBuilderFactory = new S3ClientBuilderFactory(s3SourceConfig);
+        final S3ObjectHandler s3Handler;
+        final S3ObjectRequest.Builder s3ObjectRequestBuilder = new S3ObjectRequest.Builder(buffer, s3SourceConfig.getNumberOfRecordsToAccumulate(),
+                s3SourceConfig.getBufferTimeout(), s3ObjectPluginMetrics);
+        final BiConsumer<Event, S3ObjectReference> eventMetadataModifier = new EventMetadataModifier(
+                s3SourceConfig.getMetadataRootKey());
+        if (s3SelectOptional.isPresent()) {
+            S3ObjectRequest s3ObjectRequest = s3ObjectRequestBuilder.queryStatement(s3SelectOptional.get().getQueryStatement())
+                    .serializationFormatOption(s3SelectOptional.get().getS3SelectSerializationFormatOption())
+                    .s3AsyncClient(s3ClientBuilderFactory.getS3AsyncClient()).eventConsumer(eventMetadataModifier).
+                    fileHeaderInfo(FileHeaderInfo.valueOf(s3SelectOptional.get().getCsvFileHeaderInfo().toUpperCase()))
+                    .compressionType(CompressionType.valueOf(s3SourceConfig.getCompression().name()))
+                    .s3SelectResponseHandler(new S3SelectResponseHandler()).build();
+            s3Handler = new S3SelectObjectWorker(s3ObjectRequest);
+        } else {
+            if(s3SourceConfig.getCodec() == null)
+                throw new NoSuchElementException("codec is required in pipeline yaml configuration");
+            final PluginModel codecConfiguration = s3SourceConfig.getCodec();
+            final PluginSetting codecPluginSettings = new PluginSetting(codecConfiguration.getPluginName(), codecConfiguration.getPluginSettings());
+            final Codec codec = pluginFactory.loadPlugin(Codec.class, codecPluginSettings);
+            final S3ObjectRequest s3ObjectRequest = s3ObjectRequestBuilder.bucketOwnerProvider(bucketOwnerProvider)
+                    .eventConsumer(eventMetadataModifier).codec(codec).s3Client(s3ClientBuilderFactory.getS3Client())
```

Comment:
It looks like event() is built first and then codec is applied. Is that the correct order? Is it not possible (and correct thing) to apply codec() before creating the event?